adafruit_minimqtt/adafruit_minimqtt.py

```diff
@@ -909,12 +911,21 @@ def _wait_for_msg(self, timeout=0.1):
                     "Unexpected PINGRESP returned from broker: {}.".format(sz)
                 )
             return MQTT_PINGRESP
-        if res[0] & 0xF0 != 0x30:
+
+        if res[0] & 0xF0 != MQTT_PUBLISH:
             return res[0]
+
+        # Handle only the PUBLISH packet type from now on.
         sz = self._recv_len()
         # topic length MSB & LSB
         topic_len = self._sock_exact_recv(2)
         topic_len = (topic_len[0] << 8) | topic_len[1]
+
+        if topic_len > sz - 2:
```

Comment:
What's the rationale for `sz - 2`?

Reply:
The `sz` is the "remaining length" from the MQTT fixed (sic!) header. It specifies how many bytes will follow. For the PUBLISH type, a variable header follows the fixed header. The variable header should always start with a topic name. The topic name should be [encoded as UTF-8 string](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_UTF-8_Encoded_String) which means the length will be represented as 2 bytes (MSB and LSB - as also visible in the code), unlike the variable encoding used in the fixed header for the remaining length (hence the sic).

If the 2 bytes say that the topic length is bigger than the remaining length sans the 2 bytes, the packet is internally inconsistent.

For a MQTT packet with empty message, where remaining length is R, the topic length should be equal R - 2.